applications/nrf_desktop/src/modules/led_stream.c

```diff
@@ -0,0 +1,233 @@
+{
+	LOG_INF("Send_effect to leds");
+	struct led_event *led_event = new_led_event();
+
+	led_event->stream = true;
+	led_event->led_id = led_id;
+	led_event->led_effect = effect;
+
+	__ASSERT_NO_MSG(led_event->led_effect->steps);
+
+	EVENT_SUBMIT(led_event);
+}
+
+static u8_t count_free_places(void)
+{
+	if (rx_idx > tx_idx) {
+		return QUEUE_SIZE - (rx_idx - tx_idx);
+	} else if (rx_idx < tx_idx) {
+		return QUEUE_SIZE - (tx_idx - rx_idx);
+	} else {
+		return QUEUE_SIZE;
+	}
+}
+
+static void store_data(const struct event_dyndata *data)
+{
+	u8_t free_places = count_free_places();
+
+	if (free_places > 0) {
+
+		LOG_INF("Insert data on position %" PRIu8
+		", free places %" PRIu8, rx_idx, free_places);
+		queue_data(data);
+		rx_idx++;
+		if (rx_idx == QUEUE_SIZE) {
+			rx_idx = 0;
+		}
+	} else {
+		LOG_INF("Queue is full - dropping incoming step");
+	}
+}
+
+static void send_data_from_queue(void)
+{
+	u8_t free_places = count_free_places();
+
+	if (free_places < QUEUE_SIZE) {
+		LOG_INF("Sending data: ");
+		for (size_t i = 0; i < CONFIG_DESKTOP_LED_COLOR_COUNT; i++) {
+			LOG_INF("%" PRIu8,
+				steps_queue[tx_idx].color.c[i]);
+		}
+		LOG_INF("From pos %" PRIu8 ", free places %" PRIu8,
+			tx_idx, free_places);
+
+		led_stream_effect.steps = &steps_queue[tx_idx];
+		send_effect(&led_stream_effect, STREAM_LED_ID);
+
+		tx_idx++;
+		if (tx_idx == QUEUE_SIZE) {
+			tx_idx = 0;
+		}
+	} else {
+		LOG_INF("No steps ready in queue");
+		first_sent = false;
+	}
+}
+
+static void handle_incoming_step(const struct config_event *event)
+{
+	if (leds_initialized) {
+		if (event->dyndata.size != STREAM_DATA_SIZE) {
+			LOG_WRN("Invalid stream data size (%" PRIu8 ")",
+				event->dyndata.size);
+		} else {
+			store_data(&event->dyndata);
+			if (!first_sent) {
+				LOG_INF("Sending first led effect");
+				send_data_from_queue();
+				first_sent = true;
+			}
+		}
+	} else {
+		LOG_WRN("Leds module not ready");
+	}
+}
+
+static void handle_led_ready_event(const struct led_ready_event *event)
+{
+	LOG_INF("Led ready event handler");
+	send_data_from_queue();
+}
+
+static void handle_config_event(const struct config_event *event)
+{
+	if (!event->store_needed) {
+		/* Accept only events coming from transport. */
+		return;
+	}
+
+	if (GROUP_FIELD_GET(event->id) != EVENT_GROUP_LED_STREAM) {
+		/* Only LED STREAM events. */
+		return;
+	}
+
+	switch (TYPE_FIELD_GET(event->id)) {
+	case STREAM_DATA:
+		handle_incoming_step(event);
+		break;
+
+	default:
+		LOG_WRN("Unknown LED STREAM config event");
+		break;
+	}
+}
+
+static void handle_config_fetch_request_event(
+	const struct config_fetch_request_event *event)
+{
+	LOG_INF("Handle config fetch request event");
+
+	if (GROUP_FIELD_GET(event->id) != EVENT_GROUP_LED_STREAM) {
+		/* Only LED STREAM events. */
+		return;
+	}
+
+	u8_t free_places = count_free_places();
+
+	LOG_INF("Free places left: %" PRIu8, free_places);
+
+	size_t data_size = sizeof(free_places);
```

Comment:
please do not create a temp variable for such cases

remember that each variable, function call, or sometimes even a comment is a thing you need to keep track while you analyze the code. simpler your code is easier it is to read and less bug prone :)